Controllers/HomeController.cs

```diff
@@ -3,21 +3,89 @@
 using System.Linq;
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Mvc;
+using System.Text.RegularExpressions;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using Newtonsoft.Json;
+using System.Dynamic;
+using Newtonsoft.Json.Linq;
 
 namespace Zoltu.Bags.Client.Controllers
 {
 	public class HomeController : Controller
 	{
+		private static readonly Regex productRegex = new Regex(@"product\/(\d+)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+		private static readonly Regex tagRegex = new Regex(@"tags\/(.*?)\/*?$", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+		private static readonly Regex minPriceRegex = new Regex(@"minprice\/(\d+)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+		private static readonly Regex maxPriceRegex = new Regex(@"maxprice\/(\d+)", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+		private static readonly Regex aboutRegex = new Regex(@"aboutus", RegexOptions.IgnoreCase | RegexOptions.Compiled);
+
+		// Best practices: https://www.asp.net/web-api/overview/advanced/calling-a-web-api-from-a-net-client
+		private static readonly HttpClient client = new HttpClient();
+
+		static HomeController()
+		{
+			client.BaseAddress = new Uri("https://bags-api.zoltu.com/api/");
+			client.DefaultRequestHeaders.Accept.Clear();
+			client.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
+		}
+
 		[Route("")]
 		[Route("app/{*path}")]
-		public IActionResult App(string path = null)
+		public async Task<IActionResult> App(String path = null)
 		{
-			ViewData["url"] = "https://bagcupid.com/";
-			ViewData["type"] = "website";
-			ViewData["title"] = "Bag Cupid";
-			ViewData["description"] = "What is your dream bag? Are you having trouble finding it? Let us help you!";
+			var url = "https://bagcupid.com/";
+			var type = "website";
+			var title = "Bag Cupid";
+			var description = "What is your dream bag? Are you having trouble finding it? Let us help you!";
+			var image = "https://bagcupid.com/img/logo/bagcupid.png";
+
+			if (!String.IsNullOrEmpty(path))
+			{
+				var productId = productRegex.Match(path).Groups[1].Value?.TryParseUInt64();
+				var tagId = tagRegex.Match(path).Groups[1].Value;
+				var minPrice = minPriceRegex.Match(path).Groups[1].Value.TryParseUInt64();
+				var maxPrice = maxPriceRegex.Match(path).Groups[1].Value.TryParseUInt64();
+				var aboutUs = aboutRegex.Match(path).Success;
+
+				if (productId.HasValue)
+				{
+					var product = await CallApi($"products/{productId}");
+
+					url = $"https://bagcupid.com/{path}";
+
+					if (product?.images != null && product.images.Count > 0)
+						image = (string)(((JObject)product)["images"].Aggregate((minItem, nextItem) => (int)minItem["priority"] < (int)nextItem["priority"] ? minItem : nextItem))["large"];
```

Comment:
Style: Can this be broken up into some helper functions to make the logic a little easier to follow?  Also, can we avoid some of the casting that is happening by having a strongly typed `product` object (see other comment)?  At the moment, this is a bit at risk of bugs due to the high complexity density of this code and the unchecked casts that are happening.

Also, if the chosen image doesn't have a `large` size this will either crash or end up with a `null` image (I'm not sure which).